// @qntm-code/utils/dist/cjs/dates/getters/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getToday = exports.getStartOfYear = exports.getStartOfWeek = exports.getStartOfSecond = exports.getStartOfMonth = exports.getStartOfMinute = exports.getStartOfHour = exports.getStartOfDay = exports.getEndOfYear = exports.getEndOfWeek = exports.getEndOfSecond = exports.getEndOfMonth = exports.getEndOfMinute = exports.getEndOfHour = exports.getEndOfDay = void 0;
var getEndOfDay_1 = require("./getEndOfDay");
Object.defineProperty(exports, "getEndOfDay", { enumerable: true, get: function () { return getEndOfDay_1.getEndOfDay; } });
var getEndOfHour_1 = require("./getEndOfHour");
Object.defineProperty(exports, "getEndOfHour", { enumerable: true, get: function () { return getEndOfHour_1.getEndOfHour; } });
var getEndOfMinute_1 = require("./getEndOfMinute");
Object.defineProperty(exports, "getEndOfMinute", { enumerable: true, get: function () { return getEndOfMinute_1.getEndOfMinute; } });
var getEndOfMonth_1 = require("./getEndOfMonth");
Object.defineProperty(exports, "getEndOfMonth", { enumerable: true, get: function () { return getEndOfMonth_1.getEndOfMonth; } });
var getEndOfSecond_1 = require("./getEndOfSecond");
Object.defineProperty(exports, "getEndOfSecond", { enumerable: true, get: function () { return getEndOfSecond_1.getEndOfSecond; } });
var getEndOfWeek_1 = require("./getEndOfWeek");
Object.defineProperty(exports, "getEndOfWeek", { enumerable: true, get: function () { return getEndOfWeek_1.getEndOfWeek; } });
var getEndOfYear_1 = require("./getEndOfYear");
Object.defineProperty(exports, "getEndOfYear", { enumerable: true, get: function () { return getEndOfYear_1.getEndOfYear; } });
var getStartOfDay_1 = require("./getStartOfDay");
Object.defineProperty(exports, "getStartOfDay", { enumerable: true, get: function () { return getStartOfDay_1.getStartOfDay; } });
var getStartOfHour_1 = require("./getStartOfHour");
Object.defineProperty(exports, "getStartOfHour", { enumerable: true, get: function () { return getStartOfHour_1.getStartOfHour; } });
var getStartOfMinute_1 = require("./getStartOfMinute");
Object.defineProperty(exports, "getStartOfMinute", { enumerable: true, get: function () { return getStartOfMinute_1.getStartOfMinute; } });
var getStartOfMonth_1 = require("./getStartOfMonth");
Object.defineProperty(exports, "getStartOfMonth", { enumerable: true, get: function () { return getStartOfMonth_1.getStartOfMonth; } });
var getStartOfSecond_1 = require("./getStartOfSecond");
Object.defineProperty(exports, "getStartOfSecond", { enumerable: true, get: function () { return getStartOfSecond_1.getStartOfSecond; } });
var getStartOfWeek_1 = require("./getStartOfWeek");
Object.defineProperty(exports, "getStartOfWeek", { enumerable: true, get: function () { return getStartOfWeek_1.getStartOfWeek; } });
var getStartOfYear_1 = require("./getStartOfYear");
Object.defineProperty(exports, "getStartOfYear", { enumerable: true, get: function () { return getStartOfYear_1.getStartOfYear; } });
var getToday_1 = require("./getToday");
Object.defineProperty(exports, "getToday", { enumerable: true, get: function () { return getToday_1.getToday; } });
