// @qntm-code/utils/dist/cjs/dates/setters/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setStartOfYear = exports.setStartOfWeek = exports.setStartOfSecond = exports.setStartOfMonth = exports.setStartOfMinute = exports.setStartOfHour = exports.setStartOfDay = exports.setEndOfYear = exports.setEndOfWeek = exports.setEndOfSecond = exports.setEndOfMonth = exports.setEndOfMinute = exports.setEndOfHour = exports.setEndOfDay = void 0;
var setEndOfDay_1 = require("./setEndOfDay");
Object.defineProperty(exports, "setEndOfDay", { enumerable: true, get: function () { return setEndOfDay_1.setEndOfDay; } });
var setEndOfHour_1 = require("./setEndOfHour");
Object.defineProperty(exports, "setEndOfHour", { enumerable: true, get: function () { return setEndOfHour_1.setEndOfHour; } });
var setEndOfMinute_1 = require("./setEndOfMinute");
Object.defineProperty(exports, "setEndOfMinute", { enumerable: true, get: function () { return setEndOfMinute_1.setEndOfMinute; } });
var setEndOfMonth_1 = require("./setEndOfMonth");
Object.defineProperty(exports, "setEndOfMonth", { enumerable: true, get: function () { return setEndOfMonth_1.setEndOfMonth; } });
var setEndOfSecond_1 = require("./setEndOfSecond");
Object.defineProperty(exports, "setEndOfSecond", { enumerable: true, get: function () { return setEndOfSecond_1.setEndOfSecond; } });
var setEndOfWeek_1 = require("./setEndOfWeek");
Object.defineProperty(exports, "setEndOfWeek", { enumerable: true, get: function () { return setEndOfWeek_1.setEndOfWeek; } });
var setEndOfYear_1 = require("./setEndOfYear");
Object.defineProperty(exports, "setEndOfYear", { enumerable: true, get: function () { return setEndOfYear_1.setEndOfYear; } });
var setStartOfDay_1 = require("./setStartOfDay");
Object.defineProperty(exports, "setStartOfDay", { enumerable: true, get: function () { return setStartOfDay_1.setStartOfDay; } });
var setStartOfHour_1 = require("./setStartOfHour");
Object.defineProperty(exports, "setStartOfHour", { enumerable: true, get: function () { return setStartOfHour_1.setStartOfHour; } });
var setStartOfMinute_1 = require("./setStartOfMinute");
Object.defineProperty(exports, "setStartOfMinute", { enumerable: true, get: function () { return setStartOfMinute_1.setStartOfMinute; } });
var setStartOfMonth_1 = require("./setStartOfMonth");
Object.defineProperty(exports, "setStartOfMonth", { enumerable: true, get: function () { return setStartOfMonth_1.setStartOfMonth; } });
var setStartOfSecond_1 = require("./setStartOfSecond");
Object.defineProperty(exports, "setStartOfSecond", { enumerable: true, get: function () { return setStartOfSecond_1.setStartOfSecond; } });
var setStartOfWeek_1 = require("./setStartOfWeek");
Object.defineProperty(exports, "setStartOfWeek", { enumerable: true, get: function () { return setStartOfWeek_1.setStartOfWeek; } });
var setStartOfYear_1 = require("./setStartOfYear");
Object.defineProperty(exports, "setStartOfYear", { enumerable: true, get: function () { return setStartOfYear_1.setStartOfYear; } });
