// @qntm-code/utils/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.typeOf = exports.ValueType = exports.randomNumberBetweenRange = exports.merge = exports.isString = exports.isRegExp = exports.isReactElement = exports.isPlainObject = exports.isObject = exports.isNumber = exports.isNullOrUndefined = exports.isNaNStrict = exports.isGeneratorObject = exports.isError = exports.isEqual = exports.isEmpty = exports.isDate = exports.isBuffer = exports.isArguments = exports.insertAtIndex = exports.formatTime = exports.difference = exports.clone = void 0;
__exportStar(require("./async"), exports);
var clone_1 = require("./clone");
Object.defineProperty(exports, "clone", { enumerable: true, get: function () { return clone_1.clone; } });
__exportStar(require("./dates"), exports);
var difference_1 = require("./difference");
Object.defineProperty(exports, "difference", { enumerable: true, get: function () { return difference_1.difference; } });
__exportStar(require("./dom"), exports);
var format_time_1 = require("./format-time");
Object.defineProperty(exports, "formatTime", { enumerable: true, get: function () { return format_time_1.formatTime; } });
var insertAtIndex_1 = require("./insertAtIndex");
Object.defineProperty(exports, "insertAtIndex", { enumerable: true, get: function () { return insertAtIndex_1.insertAtIndex; } });
var isArguments_1 = require("./isArguments");
Object.defineProperty(exports, "isArguments", { enumerable: true, get: function () { return isArguments_1.isArguments; } });
__exportStar(require("./isBoolean"), exports);
var isBuffer_1 = require("./isBuffer");
Object.defineProperty(exports, "isBuffer", { enumerable: true, get: function () { return isBuffer_1.isBuffer; } });
var isDate_1 = require("./isDate");
Object.defineProperty(exports, "isDate", { enumerable: true, get: function () { return isDate_1.isDate; } });
var isEmpty_1 = require("./isEmpty");
Object.defineProperty(exports, "isEmpty", { enumerable: true, get: function () { return isEmpty_1.isEmpty; } });
var isEqual_1 = require("./isEqual");
Object.defineProperty(exports, "isEqual", { enumerable: true, get: function () { return isEqual_1.isEqual; } });
var isError_1 = require("./isError");
Object.defineProperty(exports, "isError", { enumerable: true, get: function () { return isError_1.isError; } });
var isGeneratorObject_1 = require("./isGeneratorObject");
Object.defineProperty(exports, "isGeneratorObject", { enumerable: true, get: function () { return isGeneratorObject_1.isGeneratorObject; } });
var isNaNStrict_1 = require("./isNaNStrict");
Object.defineProperty(exports, "isNaNStrict", { enumerable: true, get: function () { return isNaNStrict_1.isNaNStrict; } });
var isNullOrUndefined_1 = require("./isNullOrUndefined");
Object.defineProperty(exports, "isNullOrUndefined", { enumerable: true, get: function () { return isNullOrUndefined_1.isNullOrUndefined; } });
var isNumber_1 = require("./isNumber");
Object.defineProperty(exports, "isNumber", { enumerable: true, get: function () { return isNumber_1.isNumber; } });
var isObject_1 = require("./isObject");
Object.defineProperty(exports, "isObject", { enumerable: true, get: function () { return isObject_1.isObject; } });
var isPlainObject_1 = require("./isPlainObject");
Object.defineProperty(exports, "isPlainObject", { enumerable: true, get: function () { return isPlainObject_1.isPlainObject; } });
var isReactElement_1 = require("./isReactElement");
Object.defineProperty(exports, "isReactElement", { enumerable: true, get: function () { return isReactElement_1.isReactElement; } });
var isRegExp_1 = require("./isRegExp");
Object.defineProperty(exports, "isRegExp", { enumerable: true, get: function () { return isRegExp_1.isRegExp; } });
var isString_1 = require("./isString");
Object.defineProperty(exports, "isString", { enumerable: true, get: function () { return isString_1.isString; } });
var merge_1 = require("./merge");
Object.defineProperty(exports, "merge", { enumerable: true, get: function () { return merge_1.merge; } });
var randomNumberBetweenRange_1 = require("./randomNumberBetweenRange");
Object.defineProperty(exports, "randomNumberBetweenRange", { enumerable: true, get: function () { return randomNumberBetweenRange_1.randomNumberBetweenRange; } });
var typeOf_1 = require("./typeOf");
Object.defineProperty(exports, "ValueType", { enumerable: true, get: function () { return typeOf_1.ValueType; } });
Object.defineProperty(exports, "typeOf", { enumerable: true, get: function () { return typeOf_1.typeOf; } });
__exportStar(require("./types"), exports);
